syntax = "proto3";

package grpc;
import "google/protobuf/timestamp.proto";

service Chat {
    rpc SendMessage (Msg) returns (Empty) {}
    rpc Channel (Empty) returns (stream Msg) {} /* since no terminating string */
}

service User {
    rpc GetUsersList (Empty) returns (UserList); /* Lista de usuarios */
    rpc Leave (UserID) returns (Empty); /* desconecta al usuario */
    rpc JoinServer (UserID) returns (JoinResponse); /* conexion al server */
}

service Message{
    rpc getMessages (UserID) returns (UserMsgList);
    rpc SaveUserMessage (UserID) returns (Empty);
}

message Empty{
    /* same as in google/protobuf/empty.proto */
}

message JoinResponse {
    bool response = 1;
}

message Msg {
    string id = 1;
    string message = 2;
    google.protobuf.Timestamp timestamp = 3;   
}

message UserID {
    string user_id = 1;
}

/*
Para el listado de clientes en el server de chat
*/
message UserList{
    repeated UserID users = 1;
}

/*
Para el listado de mensajes del cliente que lo solicite
*/
message UserMsgList {
    repeated Msg messages = 1;
}